src/libaktualizr/bootloader/bootloader.cc

```diff
@@ -169,13 +170,22 @@ void Bootloader::rebootFlagClear() {
 }
 
 void Bootloader::reboot(bool fake_reboot) {
+  struct stat st;
+
   if (fake_reboot) {
     boost::filesystem::remove(reboot_sentinel_);
   } else {
-    sync();
     if (setuid(0) != 0) {
       LOG_ERROR << "Failed to set/verify a root user so cannot reboot system programmatically";
+      return;
+    }
+
+    if ((stat ("/sbin/init", &st) >= 0) && (st.st_mode & (S_IXUSR | S_IXGRP | S_IXOTH))) {
+      if (system("/sbin/init 6") != 0) {
+        LOG_ERROR << "Failed to execute /sbin/init 6";
+      }
```

Comment:
Tried to figure out whether there is a POSIX way to reboot a machine, but it seems reboot or similar is not part of the POSIX spec. This StackOverflow thread suggests the same: https://unix.stackexchange.com/questions/507259/is-there-a-posix-way-to-shutdown-a-unix-machine

Is there a reason you opted for `/sbin/init 6`? Runlevel seem to be a System V init system concept, and not really standardized either,... I feel that `/sbin/reboot` is probably better suited those days, as it  clearly says what it's doing, and is supported by all init systems I know of (sysvinit, systemd, upstart)...

Reply:
The reason for choosing '/sbin/init 6' instead of '/sbin/reboot' was that: I thought it's possible that '/sbin/reboot' does not exist on some sysvinit based systems, for instance, when busybox is being used, the option 'CONFIG_INIT' is optional, which means there could be no '/sbin/reboot' compiled.

And I verified the '/sbin/init 6' worked also for systemd. (But I did not verify it for upstart)

Reply:
IMHO,
'reboot' should be part of Platform Abstraction Layer to enable platform specific implementations one of which can be specified/chosen during an image bitbaking and/or aktualizr building.
+
Default/Reference implementation targeting systemd-based system that would reboot by either calling system("systemctl reboot") or ::reboot() which effectively leads to the same effect (if you you look at the logs you can see that actually systemd takes control in case of ::reboot call and does exactly the same stuff as in case of calling systemctl reboot).

Reply:
@mike-sul 
> Default/Reference implementation targeting systemd-based system that would reboot by either calling system("systemctl reboot") or ::reboot() which effectively leads to the same effect (if you you look at the logs you can see that actually systemd takes control in case of ::reboot call and does exactly the same stuff as in case of calling systemctl reboot).

Really? How? Isn't `::reboot` calling the reboot syscall? `man 2 reboot` says that `RB_AUTOBOOT` is initiating a restart immediately, and data are lost unless synced before. How is systemd intercepting that syscall?

Reply:
Yeah, your're right. I confused it with system("reboot") which leads to the same effect as system("systemctl reboot") since the 'reboot' file/cmd is a link to systemctl 

`root@qemux86-64:/sysroot/home/root# ls -l /usr/sbin/reboot
lrwxrwxrwx   11 root     root            18 Mar 21 15:11 /usr/sbin/reboot -> /usr/bin/systemctl`

I apologize for making the incorrect statement.